File: backend/parser/lexer.py
"""Lexer that tokenizes FPB text into a stream of typed tokens."""

from dataclasses import dataclass
from typing import Optional

from parser.syntax import (
    CONNECTION_OPERATORS_SORTED,
    CONNECTION_OPERATORS,
    END_DELIMITER,
    KEYWORDS,
    PLACEMENT_ANNOTATIONS,
    START_DELIMITER,
    TokenType,
)


@dataclass
class Token:
    """A single token produced by the lexer."""
    type: TokenType
    value: str
    line: int
    column: int


class LexerError(Exception):
    """Error raised during lexical analysis."""

    def __init__(self, message: str, line: int, column: int) -> None:
        self.line = line
        self.column = column
        super().__init__(f"Line {line}, Col {column}: {message}")


class Lexer:
    """Tokenizer for FPB text syntax."""

    def __init__(self, source: str) -> None:
        self.source = source
        self.pos = 0
        self.line = 1
        self.column = 1
        self.tokens: list[Token] = []

    def tokenize(self) -> list[Token]:
        """Tokenize the entire source into a list of tokens."""
        while self.pos < len(self.source):
            self._skip_whitespace()
            if self.pos >= len(self.source):
                break

            ch = self.source[self.pos]

            if ch == "\n":
                self._emit(TokenType.NEWLINE, "\n")
                self._advance()
                continue

            if ch == "/" and self._peek(1) == "/":
                self._read_comment()
                continue

            if ch == "@":
                self._read_delimiter()
                continue

            if ch == '"':
                self._read_string()
                continue

            if ch == "{":
                self._emit(TokenType.LBRACE, "{")
                self._advance()
                continue

            if ch == "}":
                self._emit(TokenType.RBRACE, "}")
                self._advance()
                continue

            # Try connection operators
            if self._try_connection_operator():
                continue

            # Identifier or keyword
            if ch.isalpha() or ch == "_":
                self._read_word()
                continue

            # Skip unknown characters
            self._advance()

        self.tokens.append(Token(TokenType.EOF, "", self.line, self.column))
        return self.tokens

    def _advance(self) -> str:
        """Advance position by one character and return it."""
        ch = self.source[self.pos]
        self.pos += 1
        if ch == "\n":
            self.line += 1
            self.column = 1
        else:
            self.column += 1
        return ch

    def _peek(self, offset: int = 0) -> Optional[str]:
        """Peek at a character at current position + offset."""
        idx = self.pos + offset
        if idx < len(self.source):
            return self.source[idx]
        return None

    def _emit(self, token_type: TokenType, value: str) -> None:
        """Add a token to the output list."""
        self.tokens.append(Token(token_type, value, self.line, self.column))

    def _skip_whitespace(self) -> None:
        """Skip spaces and tabs (not newlines)."""
        while self.pos < len(self.source) and self.source[self.pos] in (" ", "\t", "\r"):
            self._advance()

    def _read_comment(self) -> None:
        """Read a // comment to end of line."""
        start_col = self.column
        start_line = self.line
        self.pos += 2
        self.column += 2
        start = self.pos
        while self.pos < len(self.source) and self.source[self.pos] != "\n":
            self.pos += 1
            self.column += 1
        value = "//" + self.source[start:self.pos]
        self.tokens.append(Token(TokenType.COMMENT, value.strip(), start_line, start_col))

    def _read_delimiter(self) -> None:
        """Read @startfpb, @endfpb delimiter or @annotation token."""
        start_col = self.column
        for delim, ttype in [(START_DELIMITER, TokenType.START_FPB), (END_DELIMITER, TokenType.END_FPB)]:
            if self.source[self.pos:self.pos + len(delim)] == delim:
                self._emit(ttype, delim)
                for _ in range(len(delim)):
                    self._advance()
                return
        # Check placement annotations
        for annotation in PLACEMENT_ANNOTATIONS:
            if self.source[self.pos:self.pos + len(annotation)] == annotation:
                end_pos = self.pos + len(annotation)
                if end_pos >= len(self.source) or not self.source[end_pos].isalnum():
                    self._emit(TokenType.ANNOTATION, annotation)
                    for _ in range(len(annotation)):
                        self._advance()
                    return
        # Unknown @ token - skip
        self._advance()

    def _read_string(self) -> None:
        """Read a double-quoted string literal."""
        start_col = self.column
        start_line = self.line
        self._advance()  # skip opening quote
        start = self.pos
        while self.pos < len(self.source) and self.source[self.pos] != '"' and self.source[self.pos] != "\n":
            self.pos += 1
            self.column += 1
        value = self.source[start:self.pos]
        if self.pos < len(self.source) and self.source[self.pos] == '"':
            self.pos += 1
            self.column += 1
        self.tokens.append(Token(TokenType.STRING, value, start_line, start_col))

    def _try_connection_operator(self) -> bool:
        """Try to match a connection operator at current position."""
        remaining = self.source[self.pos:]
        for op in CONNECTION_OPERATORS_SORTED:
            if remaining.startswith(op):
                token_type = CONNECTION_OPERATORS[op]
                self._emit(token_type, op)
                for _ in range(len(op)):
                    self._advance()
                return True
        return False

    def _read_word(self) -> None:
        """Read an identifier or keyword."""
        start_col = self.column
        start_line = self.line
        start = self.pos
        while self.pos < len(self.source) and (self.source[self.pos].isalnum() or self.source[self.pos] == "_"):
            self.pos += 1
            self.column += 1
        word = self.source[start:self.pos]
        if word in KEYWORDS:
            self.tokens.append(Token(TokenType.KEYWORD, word, start_line, start_col))
        else:
            self.tokens.append(Token(TokenType.IDENTIFIER, word, start_line, start_col))

File: vscode-extension/src/core/parser/lexer.ts
/** Lexer that tokenizes FPD text into a stream of typed tokens. */

import {
    CONNECTION_OPERATORS,
    CONNECTION_OPERATORS_SORTED,
    END_DELIMITER,
    KEYWORDS,
    PLACEMENT_ANNOTATIONS,
    START_DELIMITER,
    TokenType,
} from './syntax';

export interface Token {
    type: TokenType;
    value: string;
    line: number;
    column: number;
}

export class LexerError extends Error {
    line: number;
    column: number;

    constructor(message: string, line: number, column: number) {
        super(`Line ${line}, Col ${column}: ${message}`);
        this.line = line;
        this.column = column;
    }
}

export class Lexer {
    private source: string;
    private pos = 0;
    private line = 1;
    private column = 1;
    private tokens: Token[] = [];

    constructor(source: string) {
        this.source = source;
    }

    tokenize(): Token[] {
        while (this.pos < this.source.length) {
            this.skipWhitespace();
            if (this.pos >= this.source.length) {
                break;
            }

            const ch = this.source[this.pos];

            if (ch === '\n') {
                this.emit(TokenType.NEWLINE, '\n');
                this.advance();
                continue;
            }

            if (ch === '/' && this.peek(1) === '/') {
                this.readComment();
                continue;
            }

            if (ch === '@') {
                this.readDelimiter();
                continue;
            }

            if (ch === '"') {
                this.readString();
                continue;
            }

            if (ch === '{') {
                this.emit(TokenType.LBRACE, '{');
                this.advance();
                continue;
            }

            if (ch === '}') {
                this.emit(TokenType.RBRACE, '}');
                this.advance();
                continue;
            }

            if (this.tryConnectionOperator()) {
                continue;
            }

            if (this.isAlpha(ch) || ch === '_') {
                this.readWord();
                continue;
            }

            // Skip unknown characters
            this.advance();
        }

        this.tokens.push({ type: TokenType.EOF, value: '', line: this.line, column: this.column });
        return this.tokens;
    }

    private advance(): string {
        const ch = this.source[this.pos];
        this.pos++;
        if (ch === '\n') {
            this.line++;
            this.column = 1;
        } else {
            this.column++;
        }
        return ch;
    }

    private peek(offset: number = 0): string | undefined {
        const idx = this.pos + offset;
        if (idx < this.source.length) {
            return this.source[idx];
        }
        return undefined;
    }

    private emit(tokenType: TokenType, value: string): void {
        this.tokens.push({ type: tokenType, value, line: this.line, column: this.column });
    }

    private skipWhitespace(): void {
        while (this.pos < this.source.length) {
            const ch = this.source[this.pos];
            if (ch === ' ' || ch === '\t' || ch === '\r') {
                this.advance();
            } else {
                break;
            }
        }
    }

    private readComment(): void {
        const startCol = this.column;
        const startLine = this.line;
        this.pos += 2;
        this.column += 2;
        const start = this.pos;
        while (this.pos < this.source.length && this.source[this.pos] !== '\n') {
            this.pos++;
            this.column++;
        }
        const value = '//' + this.source.substring(start, this.pos);
        this.tokens.push({ type: TokenType.COMMENT, value: value.trim(), line: startLine, column: startCol });
    }

    private readDelimiter(): void {
        const delimiters: [string, TokenType][] = [
            [START_DELIMITER, TokenType.START_FPD],
            [END_DELIMITER, TokenType.END_FPD],
        ];

        for (const [delim, ttype] of delimiters) {
            if (this.source.substring(this.pos, this.pos + delim.length) === delim) {
                this.emit(ttype, delim);
                for (let i = 0; i < delim.length; i++) {
                    this.advance();
                }
                return;
            }
        }

        // Check placement annotations
        for (const annotation of PLACEMENT_ANNOTATIONS) {
            if (this.source.substring(this.pos, this.pos + annotation.length) === annotation) {
                const endPos = this.pos + annotation.length;
                if (endPos >= this.source.length || !this.isAlphaNumeric(this.source[endPos])) {
                    this.emit(TokenType.ANNOTATION, annotation);
                    for (let i = 0; i < annotation.length; i++) {
                        this.advance();
                    }
                    return;
                }
            }
        }

        // Unknown @ token - skip
        this.advance();
    }

    private readString(): void {
        const startCol = this.column;
        const startLine = this.line;
        this.advance(); // skip opening quote
        const start = this.pos;
        while (this.pos < this.source.length && this.source[this.pos] !== '"' && this.source[this.pos] !== '\n') {
            this.pos++;
            this.column++;
        }
        const value = this.source.substring(start, this.pos);
        if (this.pos < this.source.length && this.source[this.pos] === '"') {
            this.pos++;
            this.column++;
        }
        this.tokens.push({ type: TokenType.STRING, value, line: startLine, column: startCol });
    }

    private tryConnectionOperator(): boolean {
        const remaining = this.source.substring(this.pos);
        for (const op of CONNECTION_OPERATORS_SORTED) {
            if (remaining.startsWith(op)) {
                const tokenType = CONNECTION_OPERATORS.get(op)!;
                this.emit(tokenType, op);
                for (let i = 0; i < op.length; i++) {
                    this.advance();
                }
                return true;
            }
        }
        return false;
    }

    private readWord(): void {
        const startCol = this.column;
        const startLine = this.line;
        const start = this.pos;
        while (this.pos < this.source.length && (this.isAlphaNumeric(this.source[this.pos]) || this.source[this.pos] === '_')) {
            this.pos++;
            this.column++;
        }
        const word = this.source.substring(start, this.pos);
        if (KEYWORDS.has(word)) {
            this.tokens.push({ type: TokenType.KEYWORD, value: word, line: startLine, column: startCol });
        } else {
            this.tokens.push({ type: TokenType.IDENTIFIER, value: word, line: startLine, column: startCol });
        }
    }

    private isAlpha(ch: string): boolean {
        return (ch >= 'a' && ch <= 'z') || (ch >= 'A' && ch <= 'Z');
    }

    private isAlphaNumeric(ch: string): boolean {
        return this.isAlpha(ch) || (ch >= '0' && ch <= '9');
    }
}
